### PR TITLE
Add documentation and unittests regarding multisets in std.algorithm.setops

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: std/algorithm/setops.d [hunk 1]
```diff
@@ -3,6 +3,18 @@
 This is a submodule of $(MREF std, algorithm).
 It contains generic algorithms that implement set operations.
 
+The functions : $(LREF multiwayMerge), $(LREF multiwayUnion),
```
Comment (wilzbach) [inline std/algorithm/setops.d]: / :/d (remove the `:`)

File: std/algorithm/setops.d [hunk 2]
```diff
@@ -3,6 +3,18 @@
...
+$(LREF setDifference), $(LREF setIntersection), $(LREF setSymmetricDifference)
+expect a range of sorted ranges as input; providing a range of unsorted
+ranges will lead to a behaviour which is independent with the specification.
```
Comment (wilzbach) [inline std/algorithm/setops.d]: Remove this subsentence - it's expected that it's UB if you violate the contract and this doesn't read well.

File: std/algorithm/setops.d [hunk 3]
```diff
@@ -3,6 +3,18 @@
...
+
+The algorithms implemented specifically for sets : $(LREF setDifference),
+$(LREF setIntersection), $(LREF setSymmetricDifference) are not limited at
+working only with sets, but also with multisets. The rule of thumb is the following:
+if the input contains only sets, the result is going to be a set, otherwise
```
Comment (wilzbach) [inline std/algorithm/setops.d]: How about being precise here? Either all algorithms behave like this or it makes sense to make a short table.

File: std/algorithm/setops.d [hunk 4]
```diff
@@ -14,8 +26,9 @@ $(T2 largestPartialIntersectionWeighted,
         Copies out the values that occur most frequently (multiplied by
         per-value weights) in a range of ranges.)
 $(T2 multiwayMerge,
-        Computes the union of a set of sets implemented as a range of sorted
-        ranges.)
+        Computes the merging of a range of sorted ranges.)
+$(T2 multiwayUnion,
+        Computes the union of a range of sorted ranges.)
```
Comment (wilzbach) [inline std/algorithm/setops.d]: Don't forget to update the overview bookmark table at `std/algorithm/package.d`

File: std/algorithm/setops.d [hunk 5]
```diff
@@ -759,6 +813,11 @@ MultiwayMerge) is $(BIGOH n * ror.length * log(ror.length)), i.e., $(D
 log(ror.length)) times worse than just spanning all ranges in
 turn. The output comes sorted (unstably) by $(D less).
 
+The length of the resulting range is going to be in all cases
```
Comment (wilzbach) [inline std/algorithm/setops.d]: /going to be in all cases/d

File: std/algorithm/setops.d [hunk 6]
```diff
@@ -759,6 +813,11 @@ MultiwayMerge) is $(BIGOH n * ror.length * log(ror.length)), i.e., $(D
...
+the sum of all the lengths of the ranges passed as input. This
+means that all elements (duplicates or not) are transferred to
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/or not/included

File: std/algorithm/setops.d [hunk 7]
```diff
@@ -1019,6 +1117,12 @@ Lazily computes the intersection of two or more input ranges $(D
 ranges). The ranges are assumed to be sorted by $(D less). The element
 types of the ranges must have a common type.
 
+If all the ranges are sets (do not contain duplicated elements),
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/do no contain/i.e. without/

File: std/algorithm/setops.d [hunk 8]
```diff
@@ -1177,6 +1288,12 @@ r2). The two ranges are assumed to be sorted by $(D less), and the
 output is also sorted by $(D less). The element types of the two
 ranges must have a common type.
 
+If both ranges are sets (do not contain duplicate elements), the resulting
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/do no contain/i.e. without/

**Comment by RazvanN7** [comment]
@wilzbach thanks for the review. hope this is better.

> In other words: now that you have picked new symbol names, please don't expose the structs!

I'll make a PR after this is merged

File: std/algorithm/setops.d [hunk 9]
```diff
@@ -1019,6 +1115,12 @@ Lazily computes the intersection of two or more input ranges $(D
 ranges). The ranges are assumed to be sorted by $(D less). The element
 types of the ranges must have a common type.
 
+If all the ranges are sets (without duplicated elements),
+the resulting range is going to be a set, otherwise the result
+is going to be a multiset. In the case of multisets, the range with
+the minimum number of occurences of a given element, propagates the
+number of occurences of that element in the resulting range.
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/in/to/

File: std/algorithm/setops.d [hunk 10]
```diff
@@ -625,6 +658,12 @@ import std.algorithm.sorting : SortOutput; // FIXME
 Similar to $(D largestPartialIntersection), but associates a weight
 with each distinct element in the intersection.
 
+If at least one of the ranges is a multiset, then all of the occurences
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/all of the/all/

File: std/algorithm/setops.d [hunk 11]
```diff
@@ -568,6 +580,11 @@ array of the occurrences and then selecting its top items, and also
 requires less memory ($(D largestPartialIntersection) builds its
 result directly in $(D tgt) and requires no extra memory).
 
+If at least one of the ranges is a multiset, then all of the
+occurences of a duplicate element are taken into account. The result
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/all of the/all/

File: std/algorithm/setops.d [hunk 12]
```diff
@@ -568,6 +580,11 @@ array of the occurrences and then selecting its top items, and also
...
+is equivalent to merging all the ranges and picking the most
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/all the/all input/

File: std/algorithm/setops.d [hunk 13]
```diff
@@ -625,6 +658,12 @@ import std.algorithm.sorting : SortOutput; // FIXME
 Similar to $(D largestPartialIntersection), but associates a weight
 with each distinct element in the intersection.
 
+If at least one of the ranges is a multiset, then all of the occurences
+of a duplicate element are taken into account. The result
+is equivalent to merging all the ranges and picking the highest $(D tgt.length)
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/all the/all input/

Comment (wilzbach) [inline std/algorithm/setops.d]: I think you need a comma: "highest N, weight..."

File: std/algorithm/setops.d [hunk 14]
```diff
@@ -759,6 +812,10 @@ MultiwayMerge) is $(BIGOH n * ror.length * log(ror.length)), i.e., $(D
 log(ror.length)) times worse than just spanning all ranges in
 turn. The output comes sorted (unstably) by $(D less).
 
+The length of the resulting range is the sum of all the lengths of
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/of all the/of all/

File: std/algorithm/setops.d [hunk 15]
```diff
@@ -901,13 +974,28 @@ auto multiwayUnion(alias less = "a < b", RangeOfRanges)(RangeOfRanges ror)
...
+    assert(equal(multiwayUnion(b), witness));
 }
 
 /**
 Lazily computes the difference of $(D r1) and $(D r2). The two ranges
 are assumed to be sorted by $(D less). The element types of the two
 ranges must have a common type.
 
+In the case of multisets, considering that element `a` appears `x`
+times in $(D r1) and `y` times and $(D r2), the number of occurentes
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/occurentes/occurences/

File: std/algorithm/setops.d [hunk 16]
```diff
@@ -901,13 +974,28 @@ auto multiwayUnion(alias less = "a < b", RangeOfRanges)(RangeOfRanges ror)
...
+of `a` in the result range is going to be `x-y` if x > y or 0 othwerise.
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/result/resulting/

File: std/algorithm/setops.d [hunk 17]
```diff
@@ -1019,6 +1115,12 @@ Lazily computes the intersection of two or more input ranges $(D
 ranges). The ranges are assumed to be sorted by $(D less). The element
 types of the ranges must have a common type.
 
+If all the ranges are sets (without duplicated elements),
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/all the/all input/

Comment (wilzbach) [inline std/algorithm/setops.d]: I would add "i.e." to the parentheses

File: std/algorithm/setops.d (hunk 9 above)
Comment (wilzbach) [inline std/algorithm/setops.d]: s/that/this/

File: std/algorithm/setops.d [hunk 18]
```diff
@@ -1177,6 +1286,12 @@ r2). The two ranges are assumed to be sorted by $(D less), and the
 output is also sorted by $(D less). The element types of the two
 ranges must have a common type.
 
+If both ranges are sets (without duplicated elements), the resulting
+range is going to be a set. If at least one of the ranges is a multiset,
+the number of occurences of an element `x` in the resulting range is `abs(a-b)`
+where `a` is the number of occurences of `x` in $(D r1), `b` is the number of
+occurences of `x` in $(D r2) and `abs` is the absolute value.
```
Comment (wilzbach) [inline std/algorithm/setops.d]: Common D style is to use the Oxford comma: ", and"

File: std/algorithm/setops.d [hunk 19]
```diff
@@ -3,6 +3,17 @@
 This is a submodule of $(MREF std, algorithm).
 It contains generic algorithms that implement set operations.
 
+The functions $(LREF multiwayMerge), $(LREF multiwayUnion), $(LREF setDifference),
+$(LREF setIntersection), $(LREF setSymmetricDifference) expect a range of sorted
+ranges as input.
+
+All the algorithms, except (LREF multiwayMerge) and (LREF mutiwayUnion),
+are generalised to accept as input not only sets but also multisets:
```
Comment (wilzbach) [inline std/algorithm/setops.d]: I think the common style is American English -> generalized
Grep on "generalize" at dmd/phobos found three matches, none for the British variant.

File: std/algorithm/setops.d [hunk 20]
```diff
@@ -3,6 +3,17 @@
...
+when all the ranges passed as input are sets, the result is going to be
```
Comment (wilzbach) [inline std/algorithm/setops.d]: s/all the/all/

File: std/algorithm/setops.d [hunk 21]
```diff
@@ -3,6 +3,17 @@
...
+a set; if one of the ranges is a multiset, the result is going to be a multiset.
```
Comment (wilzbach) [inline std/algorithm/setops.d]: I would simply end the sentence here - the semicolon isn't a common style.

Comment (wilzbach) [inline std/algorithm/setops.d]: How about wrap this in `$(UL ..)` and then use the single points as `$(LI ...)`?

**Comment by RazvanN7** [comment]
@wilzbach I hope to get accustomed to the documentation guidelines ASAP. Till then, hope this is fine.

**Comment by wilzbach** [comment]
@CyberShadow I observed something interesting:

- https://patch-diff.githubusercontent.com/raw/dlang/phobos/pull/5627.diff shows the actual diff
- `refs/pull/5627/head` still points to the old commit

@RazvanN7 thanks a lot! I had to rebase to master due to our current GitHub problem here :/

**Comment by wilzbach** [comment]
> I had to rebase to master due to our current GitHub problem here :/

This doesn't seem to work all the time either.
@CyberShadow how we proceed here?
- Use the admin overwrite when auto-tester is green and hope that we don't accidentally break sth.?
- Try to rebase & recreate the PR and hope it works then?

**Comment by CyberShadow** [comment]
> Try to rebase & recreate the PR and hope it works then?

I think that could work, assuming you never update the branch since it's created.

We could also just wait and hope that our GitHub overlords fix the problem soon (assuming they haven't forgotten about us already)...

**Comment by wilzbach** [comment]
> I think that could work, assuming you never update the branch since it's created.

If that goes on like this, 2017H2 will have a huge PR spike :/

-> https://github.com/dlang/phobos/pull/5643

> We could also just wait and hope that our GitHub overlords fix the problem soon (assuming they haven't forgotten about us already)...

Hmm, the show must go on and I'm not too certain that our gods care much about their children.
If this state of PR redness continues, I could write a dlang-bot plugin to resubmit a rebased PR :angry:

**Pull request closed without merging**